doc/support_arbitrary_select.md

```diff
@@ -0,0 +1,62 @@
+# Design: Support arbitrary select statements
+
+SQLFlow extends the SQL syntax to enable model training and inference. The SQL extension should be natural to learn and
+integrate well with existing SQL programs despite many existing ones contain nested SELECT statements. Our SQLFlow
+users could append a TRAIN or PREDICT clause in those programs and add AI functionalities with ease.
+
+## Overview
+
+Given a long and nested select inside the extended SQL statement, SQLFlow
+
+1. splits the extended SQL statement into its select clause and the train clause.
+    1. SQL syntactically check the select clause via a particular SQL engine parser. (MySQL/Calcite/ODPS)
+    1. SQL The train clause will be syntactically checked by SQLFlow's parser.
+1. verifies all fields used in train clause are produced by select clause and have the desired types.
+1. generates Python submitter program which forwards the select clause to a particular SQL engine to fetch the training data.
+
+Please be aware that SQLFlow doesn't attempt to parse the nested select due to the difficulty in handling different syntax
+across SQL engines. Instead, it follows the UNIX's pipeline philosophy: forwarding the complexity to various SQL engines,
+while retrieves the data via unified DB API.
+
+![](figures/arbitrary-select.png)
+
+## Implementation
+
+### Modification on SQLFlow parser
+
+The SQLFlow parser is only responsible for parsing the train clause, instead of the full extended SQL statement. So the
+[`standardSelect`](https://github.com/sql-machine-learning/sqlflow/blob/158b098cfecf7b12479171b09c12b877ad3fb00b/sql/sql.y#L66-L71)
+struct will be a string.
+
+### Splitting the extended SQL
+
+SQLFlow splits the extended SQL by looking for consecutive tokens returned by the lexer. If SQLFlow finds the following consecutive tokens 
+[`TRAIN`, `IDENT`, `WITH`] or [`PREDICT`, `IDENT`, `USING`], it splits the SQL string at the beginning of the first token
```

Comment:
Should we use `TO TRAIN` to minimize the possibility that `TRAIN IDENT WITH` are all column names? It's a very tiny chance that this could happen but what we need should be a 100% sure the statement is split as we expected.

Reply:
NOTE: related PR: https://github.com/sql-machine-learning/sqlflow/pull/583

Reply:
Good point.

@wangkuiyi could you please file an issue explaining why we need to change from `TRAIN` to `TO TRAIN`? Then I will submit a PR to fix that issue.